DOC:  Add description of overflow errors

Kai-Striega opened the pull request:
As requested by @mattip in #13102.

Unlike Python, NumPy integers have fixed sizes. This can lead to
confusion when an integer overflow occurs and users expect NumPy integer
types to behave similarly to Python integers. This commit explains
integer overflow, an example and potential workarounds.

Kai-Striega committed:
DOC: Add description of overflow errorrs

Unlike Python, NumPy integers have fixed sizes. This can lead to
confusion when a integer overflow occurs and users expect NumPy integer
types to behave similarily to Python integers. This commit explains
integer overflow, an example and potential work arounds.

mattip commented on numpy/doc/basics.py:
```diff
@@ -260,6 +260,45 @@
+respectively ::
+
+    >>> np.iinfo(np.int) # Bounds of the default integer
+    iinfo(min=-9223372036854775808, max=9223372036854775807, dtype=int64)
```
"Bounds of the default integer" -> "Bounds of the default integer on 64-bit platforms"

Kai-Striega replied:
Good pick up. Although it's also not *always* correct as a C `long` on Windows only has 32-bits:

    Python 3.6 (64-bit) interactive window [PTVS 15.9.18254.1-15.0]
    >>> import platform
    >>> platform.architecture()
    ('64bit', 'WindowsPE')
    >>> import numpy as np
    >>> np.iinfo(np.int)
    iinfo(min=-2147483648, max=2147483647, dtype=int32)

maybe -> "Bounds of default int on this system"?

mattip replied:
ok

mattip commented on numpy/doc/basics.py:
```diff
@@ -260,6 +260,45 @@
+
+NumPy provides :ref: ``numpy.iinfo`` and :ref: ``numpy.finfo`` to verify the
+minimum or maximum values of NumPy integer and floating point values
+respectively ::
```
single backticks and no space between the `:ref:` and the content to make it a link

mattip commented on numpy/doc/basics.py:
```diff
@@ -260,6 +260,45 @@
+The fixed size of NumPy numeric types may cause overflow errors when a value
+requires more memory than available in the data type. For example ``np.power``
+evaluates ``100 * 10 ** 8`` correctly for 64-bit integers, but gives
+1874919424 (not correct) for a 32-bit integer.
```
single backticks around `np.power` to make it a reference

Kai-Striega replied:
If I want `np.power` to make it a reference must I also include ``:ref:``?

mattip replied:
you have to use `numpy.power`, the reference lookup doesn't know about `np`. In this case you don't need the `:ref:` since the default domain is python it will look up the object as a python reference.

mattip commented:
Thanks. The [result](https://6318-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/user/basics.types.html#overflow-errors) will provide a helpful reference for new users. Just a few small nits about backticks. Double backticks are used to render code, single backticks for references.

Kai-Striega commented:
Ok I've fixed the refs and comment. Will this be enough to close #13102?

mattip commented on numpy/doc/basics.py:
```diff
@@ -260,6 +260,45 @@
+
+NumPy provides :ref:`numpy.iinfo` and :ref:`numpy.finfo` to verify the
+minimum or maximum values of NumPy integer and floating point values
+respectively ::
```
the `numpy.power`, `numpy.iinfo` and `numpy.finfo` should **not** have `:ref:`, they should just be `single-back-tick content single-back-tick`.  I am not sure which domain is the correct one to get the links to appear, but `:ref:` apparently is not it. I tried this by doing `(cd ..; pip install .); make clean; make html` in the `doc` directory, and then `firefox build/html/user/basics.types.html` to view the result. The `pip install` is needed since the changes are part of the docstring picked up by `autodoc`, so the docstring (and therefore numpy) must be installed in the python used to build the docs

mattip replied:
> I am not sure which domain is the correct one to get the links to appear, but :ref: apparently is not it

The correct domain is `:obj:`

Kai-Striega replied:
I've removed the ``:ref:``'s entirely. That should make it simpler to read anyway. I've also rebased the commits to keep the history neat.

Kai-Striega commented:
@mattip is this ready to merge? I don't have access to the docs built during the test? Also, should #13102 be closed once this has been merged?

Kai-Striega committed:
DOC: Remove extra letter from docs

mattip commented:
[Here](https://6335-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/user/basics.types.html#overflow-errors) is the rendered page. I can see this from the circle-ci test result page by clicking on the Artifacts tab and drilling down to the correct html page (you may not be able to see the tab).

LGTM, just one tiny nit - `np.power` -> `numpy.power` so it becomes a link

Kai-Striega committed:
DOC: Minor fix in Integer Overflow doc

Change `np.power` -> `numpy.power` to make it reference
the function's documentation

Kai-Striega commented:
@mattip fixed 👍. And I'm unable to see the tab. I think you must be a member to view the artifacts, at least that's the case for SciPy.

mattip commented:
Thanks @Kai-Striega
